zerver/migrations/0373_create_role_based_system_groups.py

```diff
@@ -0,0 +1,121 @@
+    UserProfile.ROLE_REALM_ADMINISTRATOR = 200
+    UserProfile.ROLE_MODERATOR = 300
+    UserProfile.ROLE_MEMBER = 400
+    UserProfile.ROLE_GUEST = 600
+
+    realms = Realm.objects.all()
+
+    SYSTEM_USER_GROUP_ROLE_MAP = {
+        UserProfile.ROLE_REALM_OWNER: {
+            "name": "@role:owners",
+            "description": "Owners of this organization",
+        },
+        UserProfile.ROLE_REALM_ADMINISTRATOR: {
+            "name": "@role:administrators",
+            "description": "Administrators of this organization, including owners",
+        },
+        UserProfile.ROLE_MODERATOR: {
+            "name": "@role:moderators",
+            "description": "Moderators of this organization, including administrators",
+        },
+        UserProfile.ROLE_MEMBER: {
+            "name": "@role:members",
+            "description": "Members of this organization, not including guests",
+        },
+        UserProfile.ROLE_GUEST: {
+            "name": "@role:everyone",
+            "description": "Everyone in this organization, including guests",
+        },
+    }
+
+    for realm in realms:
+        role_system_groups_dict = {}
+        for role in SYSTEM_USER_GROUP_ROLE_MAP.keys():
+            user_group_params = SYSTEM_USER_GROUP_ROLE_MAP[role]
+            user_group = UserGroup(
+                name=user_group_params["name"],
+                description=user_group_params["description"],
+                realm=realm,
+                is_system_group=True,
+            )
+            role_system_groups_dict[role] = user_group
+
+        full_members_system_group = UserGroup(
+            name="@role:fullmembers",
+            description="Members of this organization, not including new accounts and guests",
+            realm=realm,
+            is_system_group=True,
+        )
+        everyone_on_internet_system_group = UserGroup(
+            name="@role:internet",
+            description="Everyone on the Internet",
+            realm=realm,
+            is_system_group=True,
+        )
+
+        system_user_groups_list = [
+            role_system_groups_dict[UserProfile.ROLE_REALM_OWNER],
+            role_system_groups_dict[UserProfile.ROLE_REALM_ADMINISTRATOR],
+            role_system_groups_dict[UserProfile.ROLE_MODERATOR],
+            full_members_system_group,
+            role_system_groups_dict[UserProfile.ROLE_MEMBER],
+            role_system_groups_dict[UserProfile.ROLE_GUEST],
+            everyone_on_internet_system_group,
+        ]
+
+        UserGroup.objects.bulk_create(system_user_groups_list)
+
+        subgroup_objects = []
+        for idx, supergroup in enumerate(system_user_groups_list[1:]):
+            subgroup_objects.append(
+                GroupGroupMembership(subgroup=system_user_groups_list[idx], supergroup=supergroup)
+            )
+
+        GroupGroupMembership.objects.bulk_create(subgroup_objects)
+
+        users = UserProfile.objects.filter(realm=realm)
```

Comment:
Not sure if it's a sufficiently relevant difference to make it worth, but perhaps we can just fetch `.values('id', 'role')` here since that's all that's needed rather than fetching entire UserProfiles, since this is potentially a sizeable query? Thoughts? @timabbott

Reply:
Yeah, that's a good idea.

Reply:
Applied this change:

```
diff --git a/zerver/migrations/0375_create_role_based_system_groups.py b/zerver/migrations/0375_create_role_based_system_groups.py
index 8a26c9231a..aee7b38f72 100644
--- a/zerver/migrations/0375_create_role_based_system_groups.py
+++ b/zerver/migrations/0375_create_role_based_system_groups.py
@@ -99,7 +99,7 @@ def create_role_based_system_groups(apps: StateApps, schema_editor: DatabaseSche
 
             GroupGroupMembership.objects.bulk_create(subgroup_objects)
 
-            users = UserProfile.objects.filter(realm=realm)
+            users = UserProfile.objects.filter(realm=realm).only("id", "role", "date_joined")
             group_membership_objects = []
             for user in users:
                 system_group = role_system_groups_dict[user.role]
@@ -118,6 +118,16 @@ def create_role_based_system_groups(apps: StateApps, schema_editor: DatabaseSche
             UserGroupMembership.objects.bulk_create(group_membership_objects)
 
 
+def delete_role_based_system_groups(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    UserGroup = apps.get_model("zerver", "UserGroup")
+    GroupGroupMembership = apps.get_model("zerver", "GroupGroupMembership")
+    UserGroupMembership = apps.get_model("zerver", "UserGroupMembership")
+    with transaction.atomic():
+        GroupGroupMembership.objects.all().delete()
+        UserGroupMembership.objects.filter(user_group__is_system_group=True).delete()
+        UserGroup.objects.filter(is_system_group=True).delete()
+
+
 class Migration(migrations.Migration):
     atomic = False
 
@@ -127,6 +137,8 @@ class Migration(migrations.Migration):
 
     operations = [
         migrations.RunPython(
-            create_role_based_system_groups, reverse_code=migrations.RunPython.noop, elidable=True
+            create_role_based_system_groups,
+            reverse_code=delete_role_based_system_groups,
+            elidable=True,
         ),
     ]
```